// @n8n/client-oauth2/dist/CredentialsFlow.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CredentialsFlow = void 0;
const constants_1 = require("./constants");
const utils_1 = require("./utils");
class CredentialsFlow {
    constructor(client) {
        this.client = client;
    }
    async getToken(opts) {
        var _a;
        const options = { ...this.client.options, ...opts };
        (0, utils_1.expects)(options, 'clientId', 'clientSecret', 'accessTokenUri');
        const body = {
            grant_type: 'client_credentials',
        };
        if (options.scopes !== undefined) {
            body.scope = options.scopes.join((_a = options.scopesSeparator) !== null && _a !== void 0 ? _a : ' ');
        }
        const requestOptions = (0, utils_1.getRequestOptions)({
            url: options.accessTokenUri,
            method: 'POST',
            headers: {
                ...constants_1.DEFAULT_HEADERS,
                Authorization: (0, utils_1.auth)(options.clientId, options.clientSecret),
            },
            body,
        }, options);
        const responseData = await this.client.request(requestOptions);
        return this.client.createToken(responseData);
    }
}
exports.CredentialsFlow = CredentialsFlow;
